### PR TITLE
v8 - allows audit trail/logging of changed cultures

### PR DESCRIPTION
**Pull request by Shazwazza**
This is quite a hefty PR since there was a lot of changes/bugs required to make this happen

This resolves: #3334, #3333, https://github.com/umbraco/Umbraco.Private/issues/237

Several things have had to change:

* Need to be able to change track the culture names so there's a new CultureNameCollection
* Enhanced the ObservableDictionary so it can more easily be used everywhere (we should also update the PropertyTypeCollection and PropertyGroupCollection to use this to save on some code duplication)
* There is issues with DeepCloning when models are dealing with observable collections + events. The problem is all event handlers will remain when cloning but they will be attached to the original entity which means if the clone changes, the event is handled on the original entity which of course causes problems. This is also a problem in v7 but in some odd ways, it's not causing problems because some things are working by fluke (i.e. checking if property types should be removed in the contenttype base repo). These are all fixed now.
* The audit trail functionality needed updating so there are a couple more columns on that table. The audit trail in the UI is translatable which means that any 'comment' that goes into the database isn't actually displayed so we can't 'just' add a log in there and have it displayed in the UI, it needs to be translatable. However we then also need to inject translatable parameters into the audit trail UI which we never had before so now we have a `parameters` column in the log table for this reason. I had to add a few more audit trail types too.
* The English audit trail language values have been cleaned up to make sense, for some reason they all said "performed by user", well of course they were performed by the user. Similarly in the actually comment in the DB it said this, so this is all cleaned up to make sense and I've added a column to define which entity type the log entry was for, this now means it's much leaner as far as what data is stored.
* I've added a bunch of unit tests
* I've fixed the audit trail not showing when you switch languages in the back office when viewing the Info tab
* Other code cleanup

Testing:

* [x] Make sure that when you save a variant that the audit trail shows which one was saved, also do this for publishing. If you save or publish multiple variants at once, it will show a comma delimited string of them (see image below)
* [x] Make sure all unit tests pass, except for the new `Can_Rollback_Version_On_Multilingual`, i cannot get that one to pass but i think that is caused by a recent change in temp8
* [ ] Review the code ... sorry there is a lot

### CODE REVIEW

**Comment by warrenbuckley** [comment]
@Shazwazza I got all unit tests to pass including `Can_Rollback_Version_On_Multilingual`

**Comment by warrenbuckley** [comment]
Tested and it works as expected. I jsut need @clausjensen or @zpqrtbnk to give me a hand with the code change review as I am not sure I fully follow all the changes made.

So once the code has been sanity checked its good to be merged in !

**Comment by zpqrtbnk** [comment]
Reviewing code. Will of course want to changes things. Like, `IContent.CultureNames` does not return names anymore but some culture infos, including a date, so it should probably be renamed. Minor things. WIP.

**Comment by zpqrtbnk** [comment]
Reviewed, cleaned up, all tests green + now happy w/code. Ok to merge.
But.
Will first review the conflicts.

**Comment by zpqrtbnk** [comment]
Conflicts reviewed, ok for me to merge

File: src/Umbraco.Core/Collections/ObservableDictionary.cs [hunk 1]
```diff
@@ -155,6 +156,75 @@ public virtual void ChangeKey(TKey currentKey, TKey newKey)
...
+        }
+
+        bool ICollection<KeyValuePair<TKey, TValue>>.Contains(KeyValuePair<TKey, TValue> item)
+        {
+            return ContainsKey(item.Key);
+        }
+
+        void ICollection<KeyValuePair<TKey, TValue>>.CopyTo(KeyValuePair<TKey, TValue>[] array, int arrayIndex)
+        {
+            throw new NotImplementedException();
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Core/Collections/ObservableDictionary.cs]: Yup

**Pull request merged**